sti-kbproxy/src/main/java/uk/ac/shef/dcs/kbproxy/sparql/SPARQLProxy.java

```diff
@@ -215,69 +215,79 @@ protected boolean ask(Query sparqlQuery) {
     }
     return out;
   }
 
   protected List<String> queryForLabel(Query sparqlQuery, String resourceURI) throws KBProxyException {
-    try {
-      log.info("Sparql query: " + sparqlQuery.toString());
+    // Query all labels of the resource.
+    List<String> labels = queryReturnSingleValues(sparqlQuery, SPARQL_VARIABLE_OBJECT);
+    labels = labels.stream().filter(item -> !isNullOrEmpty(item)).collect(Collectors.toList());
+
+    // The resource has no statement with label property, apply simple heuristics to parse the
+    // resource URI.
+    if (labels.size() == 0) {
+      // URI like https://www.w3.org/1999/02/22-rdf-syntax-ns#type
+      int trimPosition = resourceURI.lastIndexOf("#");
+
+      // URI like http://dbpedia.org/property/name
+      if (trimPosition == -1) {
+        trimPosition = resourceURI.lastIndexOf("/");
+      }
 
-      QueryExecution qExec = QueryExecutionFactory.sparqlService(kbDefinition.getSparqlEndpoint(), sparqlQuery);
+      if (trimPosition != -1) {
+        // Remove anything that is not a character or digit
+        // TODO: For a future improvement, take into account the "_" character.
+        String stringValue = resourceURI.substring(trimPosition + 1).replaceAll("[^a-zA-Z0-9]", "").trim();
 
-      List<String> out = new ArrayList<>();
-      ResultSet resultSet = qExec.execSelect();
+        // Derived KBs can have custom URI conventions.
+        stringValue = applyCustomUriHeuristics(resourceURI, stringValue);
+        stringValue = StringUtils.splitCamelCase(stringValue);
 
-      resultSet.forEachRemaining(solution -> {
-        RDFNode labelNode = solution.get(SPARQL_VARIABLE_OBJECT);
-        if (labelNode == null) return;
+        labels.add(stringValue);
+      }
+    }
 
-        String label = labelNode.toString();
+    // Remove any language tags
+    String suffix = kbDefinition.getLanguageSuffix();
+    if (!isNullOrEmpty(suffix)) {
+      List<String> filteredLabels = new ArrayList<>();
 
-        if (label != null) {
-          out.add(label);
-        }
-      });
-
-      if (out.size() == 0) { //the resource has no statement with prop "rdfs:label", apply heuristics to parse the
-        //resource uri
-        int trim = resourceURI.lastIndexOf("#");
-        if (trim == -1)
-          trim = resourceURI.lastIndexOf("/");
-        if (trim != -1) {
-          String stringValue = resourceURI.substring(trim + 1).replaceAll("[^a-zA-Z0-9]", "").trim();
-          if (resourceURI.contains("yago")) { //this is an yago resource, which may have numbered ids as suffix
-            //e.g., City015467
-            int end = 0;
-            for (int i = 0; i < stringValue.length(); i++) {
-              if (Character.isDigit(stringValue.charAt(i))) {
-                end = i;
-                break;
-              }
-            }
-            if (end > 0)
-              stringValue = stringValue.substring(0, end);
+      for(String label : labels) {
+        if (label.contains("@")) {
```

Comment:
This is not the best way to get literal without the language element - when there is "@" in the literal, e.g. email address, it will not work well. Please mark that or better use Jena to parse plain literal. (without language tag)